metrics/release_test.go

```diff
@@ -0,0 +1,221 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	"github.com/redhat-appstudio/release-service/kcp"
+	"sigs.k8s.io/controller-runtime/pkg/metrics"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/testutil"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var _ = Describe("Metrics Release", Ordered, func() {
+
+	BeforeAll(func() {
+		// We need to unregister in advance otherwise it breaks with `AlreadyRegisteredError`
+		metrics.Registry.Unregister(ReleaseAttemptRunningSeconds)
+		metrics.Registry.Unregister(ReleaseAttemptConcurrentTotal)
+		metrics.Registry.Unregister(ReleaseAttemptDurationSeconds)
+	})
+
+	var (
+		AttemptRunningSecondsHeader = inputHeader{
+			Name: "release_attempt_running_seconds",
+			Help: "Release durations from the moment the release resource was created til the release is marked as running",
+		}
+		AttemptDurationSecondsHeader = inputHeader{
+			Name: "release_attempt_duration_seconds",
+			Help: "Release durations from the moment the release PipelineRun was created til the release is marked as finished",
+		}
+		AttemptTotalHeader = inputHeader{
+			Name: "release_attempt_total",
+			Help: "Total number of releases processed by the operator",
+		}
+	)
+
+	const (
+		validReleaseReason   = "valid_release_reason"
+		invalidReleaseReason = "invalid_release_reason"
+		strategy             = "nostrategy"
+	)
+
+	var defaultNamespace = kcp.NamespaceReference{
+		Namespace: "default",
+		Workspace: "default",
+	}
+
+	Context("When RegisterNewRelease is called", func() {
+
+		// As we need to share metrics within the Context, we need to use "per Context" `Before`|`After`All()
+		BeforeAll(func() {
+			// Overriding the metric definitions to simplify the tests
+			//
+			// `Help` can't be overriden due to `https://github.com/prometheus/client_golang/blob/83d56b1144a0c2eb10d399e7abbae3333bebc463/prometheus/registry.go#L314`
+			ReleaseAttemptRunningSeconds = prometheus.NewHistogram(
+				prometheus.HistogramOpts{
+					Name:    "release_attempt_running_seconds",
+					Help:    "Release durations from the moment the release resource was created til the release is marked as running",
+					Buckets: []float64{1, 5, 10, 30},
+				},
+			)
+			ReleaseAttemptConcurrentTotal = prometheus.NewGauge(
+				prometheus.GaugeOpts{
+					Name: "release_attempt_concurrent_requests",
+					Help: "Total number of concurrent release attempts",
+				},
+			)
+			metrics.Registry.MustRegister(ReleaseAttemptRunningSeconds,
```

Comment:
Same line as it's not long. Otherwise, I would write it as:
```
metrics.Registry.MustRegister(ReleaseAttemptRunningSeconds,
 
Add a suggestion, <Ctrl+g>
Add heading textAdd bold text, <Ctrl+b>Add italic text, <Ctrl+i>
Add a quote, <Ctrl+Shift+.>Add code, <Ctrl+e>Add a link, <Ctrl+k>
Add a bulleted list, <Ctrl+Shift+8>Add a numbered list, <Ctrl+Shift+7>Add a task list, <Ctrl+Shift+l>
Directly mention a user or team
Reference an issue, pull request, or discussion
Add saved reply
Same line as it's not long. Otherwise, I would write it as:
```

```
metrics.Registry.MustRegister(
        ReleaseAttemptRunningSeconds,
        ReleaseAttemptConcurrentTotal
)
```

Reply:
I guess something messed up with your paste. Can you edit?

Reply:
TF xD

I didn't see this...

Put everything in the same like as it's not long at all.
```
metrics.Registry.MustRegister(ReleaseAttemptRunningSeconds, ReleaseAttemptConcurrentTotal)
```

Otherwise, write it like:
```
metrics.Registry.MustRegister(
        ReleaseAttemptRunningSeconds,
        ReleaseAttemptConcurrentTotal
)
```